game-core/src/main/java/games/strategy/triplea/ResourceLoader.java

```diff
@@ -257,4 +260,19 @@ public boolean hasPath(final String path) {
     return UrlStreams.openStream(url)
         .orElseThrow(() -> new IllegalStateException("Failed to open an input stream to: " + path));
   }
+
+  /**
+   * Reads the given file as an image. Path is relative to the 'project root'.
+   *
+   * @param path Location of the file to be read as an image.
+   */
+  public Image getImage(final File path) {
+    checkArgument(path.exists(), "File must exist at path: " + path.getAbsolutePath());
+    checkArgument(path.isFile(), "Must be a file: " + path.getAbsolutePath());
+    try {
+      return ImageIO.read(path);
+    } catch (final IOException e) {
+      throw new RuntimeException("Unable to load image at path: " + path.getAbsolutePath(), e);
+    }
+  }
```

Comment:
Interesting consideration. My thinking was ResourceLoader controls access, at least it has the reference to the 'assets' folder and ImageFactory was not appropriate.
- subclassing is bad way to re-use code. At the risk of sounding religious, it's evil and needs to be used very judiciously.
- ImageFactory brings in a lot of unnecessary code and constraints. Caching the image is not needed and perhaps not desired. ImageFactory is documented as wanting to be called from the EDT, that is not desired for the case of unit scroller where the images can be fetched off the EDT thread.

Encapsulating access of resourceloader through ImageFactory to simply encapsulate the access seems like unnecessary indirection IMO. To boot though, resource loader is used directly by other classes.

In essence all that is really needed is a non-throwing version of ImageIO and the reference to the assets folder. Ideally this code would not be stateful, but we have the reference to assets folder that is being borrowed.

Looking for some sort of better option, I think simply moving the image loading code to its own utility class is not a bad choice: 8f5cb327fd279e0e48619ab4c9ba6f494f975d83

ImageFactory has another problem where it wants URL paths. That is a small problem as we can easily create `File` paths, creating a URL path to a file system entity is less ideal.

Reply:
Well sub-classing has its own use cases where it is superior in comparison to other ways to layout code.
That being said I don't think the whole image factory code uses object orientation very well and would benefit heavily from composition because that'd allow better unit testing because the filesystem wouldn't be required for all the "subclasses" that currently exist.
Also I think the factory class is currently misused in a lot of cases:
Because ImageIO#read is a blocking operation (at least to my knowledge), all of the UI code that tries to add a button with an `ImageIcon` will load the Images on the UI Thread.
I believe IO operations like this in general to be the main reason why starting the game causes the UI to freeze for a couple of seconds. I tried fixing this by using futures, but this really didn't work too well.

Reply:
Thinking about this I should probably just created a subclass of Button that takes a URL or InputStream and loads the actual icon in a different Thread to later add it back to itself.

Reply:
Large images could have a noticeable re-rendering. I was more thinking a background thread launched after the application is launched. There is a lot of 'dead-time' in the selection menus when loading a game where the game could be pre-fetching images that will be used for sure.